### PR TITLE
surface bidentate thermo fix

### PR DESCRIPTION
**Pull request by sevyharris**
### Motivation or Problem
Surface bidentate reaction crashed (#2039) trying to get the thermo model for some species. It looks like it sometimes tried the wrong configuration of the gas-phase molecule.

### Description of Changes
Added code to try multiple configurations and pick the one that works.
If multiple configurations are valid it uses the lowest energy one.

### Testing
Currently has thermo test for `N(=X)N(X)OH`
and tests for the molecules mentioned in the issue 2039 thread:
`C*O*` becomes `C[-1]#O[+1]`
`O*O*` becomes `[O][O]`
`C*C*` becomes `[C]#[C]`
`N*N*O` becomes `N#N=O` (the middle Nitrogen has valence 5)

### CODE REVIEW

**Comment by rwest** [comment]
Sorry I didn't have time to advise you of this earlier, but the generating the thermo of all resonance structures (all `Molecules` in a `Species`) and adopting the lowest energy one might occur later anyway, when generating the thermo of a species. It should. Though, come to think of it, it might not handle adsorbates properly yet. So maybe we do need what you did. 🤔  I need sleep

**Comment by sevyharris** [comment]
I'm not exactly sure what the species C*O* in "C*O* becomes C[-1]#O[+1]" is supposed to be. Using the rmg molecule search I narrowed it down to either:

[Pt][C-]=[O+][Pt]
1 O u0 p1 c+1 {2,D} {4,S}
2 C u0 p1 c-1 {1,D} {3,S}
3 X u0 p0 c0 {2,S}
4 X u0 p0 c0 {1,S}

or

[Pt][C]O[Pt]
1 O u0 p2 c0 {2,S} {4,S}
2 C u0 p1 c0 {1,S} {3,S}
3 X u0 p0 c0 {2,S}
4 X u0 p0 c0 {1,S}

The database says [Pt][C-]=[O+][Pt] is a representative structure, so I'll proceed with this one in the unit test unless someone corrects me.

**Comment by sevyharris** [comment]
I'm not sure if the case `N*N*O` becomes `N#N=O` is the same as case for the following adsorbate for which there is already a unit test (test_adsorbate_thermo_generation_bidentate_asymmetric_NNOH)

N(=X)N(X)OH
1 O u0 p2 c0 {2,S} {4,S}
2 N u0 p1 c0 {1,S} {3,S} {5,S}
3 N u0 p1 c0 {2,S} {6,D}
4 H u0 p0 c0 {1,S}
5 X u0 p0 c0 {2,S}
6 X u0 p0 c0 {3,D}

Would adding a separate test case be redundant here?

**Comment by josefedupe** [comment]
Hello everyone. A similar case was handled in #2039, there the error was referenced to be:

```
`For reaction generation 1 process is used.
Generating thermo for new species...
Error: Could not update atomtypes for this molecule:
1 O u0 p1 c+1 {2,T}
2 O u0 p2 c-1 {1,T}

Error: Error attempting to get thermo for species [Pt]OO[Pt](18) with structure
1 O u0 p2 c0 {2,S} {3,S}
2 O u0 p2 c0 {1,S} {4,S}
3 X u0 p0 c0 {1,S}
4 X u0 p0 c0 {2,S}`
```

the final line of the report was:

`rmgpy.exceptions.AtomTypeError: Unable to determine atom type for atom O-, which has 0 single bonds, 0 double bonds (0 to O, 0 to S, 0 others), 1 triple bonds, 0 quadruple bonds, 0 benzene bonds, 2 lone pairs, and -1 charge.`

After changing the lines presented in 08afb41, the case run correctly without errors. The final log was:

```
---------------------------------
Added 0 new core species
Created 0 new edge species
Added 0 new core reactions
Created 0 new edge reactions

After model enlargement:
    The model core has 44 species and 256 reactions
    The model edge has 167 species and 452 reactions


Adding reaction library Surface/Deutschmann_Ni_full to output file...
Saving current model core to Chemkin file...
Chemkin file contains 0 reactions.
Chemkin file contains 259 reactions.
Saving annotated version of Chemkin files...
Chemkin file contains 0 reactions.
Chemkin file contains 259 reactions.
Saving current model core and edge to Chemkin file...
Chemkin file contains 0 reactions.
Chemkin file contains 708 reactions.
Saving annotated version of Chemkin files...
Chemkin file contains 0 reactions.
Chemkin file contains 708 reactions.
Saving current model core to HTML file...
Saving current model edge to HTML file...
Updating RMG execution statistics...
    Execution time (DD:HH:MM:SS): 00:00:04:08
    Memory used: 382.79 MB

Making seed mechanism...
Performing final model checks...
No collision rate violators found in the model's core.

MODEL GENERATION COMPLETED

The final model core has 44 species and 256 reactions
The final model edge has 167 species and 452 reactions

RMG execution terminated at Tue Nov 24 23:28:39 2020`
```

The input.py file was,

```
`database(
    thermoLibraries=['surfaceThermoNi111', 'primaryThermoLibrary', 'thermo_DFT_CCSDTF12_BAC'], 
    reactionLibraries = [('Surface/Deutschmann_Ni_full', True)],
    seedMechanisms = [],
    kineticsDepositories = ['training'],
    kineticsFamilies = [
    'surface'
                       ],
    kineticsEstimator = 'rate rules',
)

catalystProperties(
    bindingEnergies = {  # values for Ni(111)
                        'H': (-2.892, 'eV/molecule'),
                        'O': (-4.989, 'eV/molecule'),
                        'C': (-6.798, 'eV/molecule'),
                        'N': (-5.164, 'eV/molecule'), 
                      },
    surfaceSiteDensity=(2.1052e-8, 'mol/cm^2'), # values for Ni(111)
)

species(
   label='O2',
   reactive=True,
   structure=SMILES("[O][O]"),
)

species(
    label='C2H6',
    reactive=True,
    structure=SMILES("CC"),
)

species(
    label='C2H4',
    reactive=True,
    structure=SMILES("C=C"),
)

species(
    label='CO',
    reactive=True,
    structure=SMILES("[C-]#[O+]"),
)

species(
    label='CO2',
    reactive=True,
    structure=SMILES("O=C=O"),
)

species(
    label='H2O',
    reactive=True,
    structure=SMILES("O"),
)


#-------

species(
    label='site',
    reactive=True,
    structure=adjacencyList("1 X u0"),
)
#----------

# Reaction systems
surfaceReactor(
    temperature=(623,'K'),
    initialPressure=(1.0, 'bar'),
    initialGasMoleFractions={
        "C2H6": 2.0,
        "C2H4": 0.0,
        "CO": 0.0,
        "CO2": 0.0,
        "O2": 1.0,
        "H2O": 0.0,
    },
    initialSurfaceCoverages={
        "site": 1.0,
    },
    surfaceVolumeRatio=(1.0e5, 'm^-1'),
    terminationConversion = { "O2":0.99,},
    terminationTime=(60.0, 's'),
)

simulator(
    atol=1e-15,
    rtol=1e-4,
)

model(
    toleranceKeepInEdge=0.0,
    toleranceMoveToCore=1e-6,
    toleranceInterruptSimulation=1e-6,
    maximumEdgeSpecies=100000
)

options(
    units='si',
    generateOutputHTML=True,
    generatePlots=False, # Enable to make plots of core and edge size etc.. But takes a lot of the total runtime!
    saveEdgeSpecies=True,
    saveSimulationProfiles=True,
    verboseComments=True,
)

# Species Constrains:
generatedSpeciesConstraints(
    maximumCarbonAtoms=2,
    allowSingletO2 = True,
)`
```

Thanks to all for the help!

**Comment by Tingchenlee** [comment]
I think [dd5171c](https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2045/commits/dd5171c6bb5aea0a649e443e022e31a38c20dae1) solve the thermo error of these species now: 
```
Error: Could not update atomtypes for this molecule:
multiplicity 1
1 N u1 p0 c0 {2,S} {3,T}
2 N u0 p1 c0 {1,S} {4,D}
3 N u2 p0 c0 {1,T}
4 N u0 p1 c0 {2,D} {5,S}
5 H u0 p0 c0 {4,S}

Error: Error attempting to get thermo for species N=NN(N=[Pt])[Pt](43) with structure 
1 N u0 p1 c0 {2,S} {3,S} {6,S}
2 N u0 p1 c0 {1,S} {4,D}
3 N u0 p1 c0 {1,S} {7,D}
4 N u0 p1 c0 {2,D} {5,S}
5 H u0 p0 c0 {4,S}
6 X u0 p0 c0 {1,S}
7 X u0 p0 c0 {3,D} 
```
and
```
Error: Could not update atomtypes for this molecule:
multiplicity 1
1 O u0 p2 c0 {2,S} {4,S}
2 N u1 p0 c0 {1,S} {3,T}
3 N u2 p0 c0 {2,T}
4 H u0 p0 c0 {1,S}

Error: Error attempting to get thermo for species ON(N=[Pt])[Pt](48) with structure 
1 O u0 p2 c0 {2,S} {4,S}
2 N u0 p1 c0 {1,S} {3,S} {5,S}
3 N u0 p1 c0 {2,S} {6,D}
4 H u0 p0 c0 {1,S}
5 X u0 p0 c0 {2,S}
6 X u0 p0 c0 {3,D} 
```
And I don't have any other thermo error at this time.

**Comment by rwest** [comment]
Looks good to me.

**Review by rwest — APPROVED** [review]
Looks good to me (but I helped write it).

File: rmgpy/data/thermo.py [hunk 1]
```diff
@@ -1600,18 +1604,42 @@ def get_thermo_data_for_surface_species(self, species):
...
+            dummy_species = Species()
+            dummy_species.molecule = [dummy_molecule]
+            dummy_species.generate_resonance_structures()
+            dummy_species.thermo = self.get_thermo_data(dummy_species)
+            gas_phase_species.append(dummy_species)
+            
+        # define the comparison function to find the lowest energy
+        def lowest_energy(species):
+            if hasattr(species.thermo, 'H298'):
+                return species.thermo.H298.value
```
Comment (davidfarinajr) [inline rmgpy/data/thermo.py:1635]: Can you please change this to `.value_si`?  `get_enthalpy` will return SI units (J/mol), but the ThermoData H298 may have different units.

**Review by davidfarinajr — CHANGES_REQUESTED** [review]
Thanks for this contribution!  Looks great!  Only change I requested was changing `value` to `value_si` to make sure we are comparing the same units for enthalpy.  Once we get that fixed, I think it's ready for merging.

File: rmgpy/data/thermo.py (hunk 1 above)
Comment (rwest) [reply rmgpy/data/thermo.py:1635]: Good catch!  I almost merged this, thinking it had already been reviewed enough. But following the best practice of rigorous peer review saves the day, again.
Thanks @davidfarinajr for doing it carefully.

Comment (rwest) [reply rmgpy/data/thermo.py:1635]:
```suggestion
                return species.thermo.H298.value_si
```

**Comment by rwest** [comment]
Huh. I made @davidfarinajr's suggested change from `.value` to `.value_si` using the inline editor in github, because it seemed like a trivial fix. And then the unit test suite failed!   So there's more work to be done ☹️

**Comment by rwest** [comment]
Running the tests locally, I get
```
======================================================================
FAIL: Test thermo generation for a bidentate adsorbate [C]#[C]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/rwest/Code/Cat/RMG-Py/rmgpy/data/thermoTest.py", line 765, in test_adsorbate_thermo_generation_bidentate_C2
    self.assertEqual(thermo.label, 'C2(T)XX', 'Should have found triplet C2 in the gas phase library')
AssertionError: '' != 'C2(T)XX'
```
So perhaps it's now picking different lowest energy electronic structure for C2 now? One from an estimate not a library (hence no label)?  Probably need to dig in to this with a debugger, to see what structures it's choosing between, what enthalpies it's using, and what it's picking.

**Comment by davidfarinajr** [comment]
> Running the tests locally, I get
> 
> ```
> ======================================================================
> FAIL: Test thermo generation for a bidentate adsorbate [C]#[C]
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/Users/rwest/Code/Cat/RMG-Py/rmgpy/data/thermoTest.py", line 765, in test_adsorbate_thermo_generation_bidentate_C2
>     self.assertEqual(thermo.label, 'C2(T)XX', 'Should have found triplet C2 in the gas phase library')
> AssertionError: '' != 'C2(T)XX'
> ```
> 
> So perhaps it's now picking different lowest energy electronic structure for C2 now? One from an estimate not a library (hence no label)? Probably need to dig in to this with a debugger, to see what structures it's choosing between, what enthalpies it's using, and what it's picking.

It is now picking `:C=C:` with group additivity enthalpy of `184 kcal/mol` https://rmg.mit.edu/database/thermo/molecule/1%20C%20u0%20p1%20c0%20%7B2,D%7D%0A2%20C%20u0%20p1%20c0%20%7B1,D%7D%0A vs `[C]#[C]` with library enthalpy of `199 kcal/mol`. This `value_si` now picks the former since the units are in `kJ/mol` (625 kJ/mol)

**Comment by davidfarinajr** [comment]
ATcT actually has singlet C2 slightly lower in energy than C2 triplet (827 kJ/mol for triplet and 820 kJ/mol for singlet).  Although the group additivity estimate is way off (625 kJ/mol for singlet).  Perhaps we should rank the thermo based on where is comes from?  Choose the lowest E from library rather than lowest E overall since group additivity is not so good for small things like C2, and very poor for small nitrogen containing things

**Comment by rwest** [comment]
I think your suggestion is probably the right one. So feel free to add a commit to this PR that implements it. 
I would still like to discuss it at the next RMG-cat meeting, to see if anyone thinks of something we've overlooked.

**Pull request merged**